src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs

```diff
@@ -116,30 +154,32 @@ public void Dispose()
             _defaultCapabilitySemaphore = null;
             _metadataSemaphore?.Dispose();
             _metadataSemaphore = null;
+            _runRebuilder = false;
         }
 
         public async Task Handle(RebuildCapabilityStatement notification, CancellationToken cancellationToken)
         {
             _logger.LogInformation("SystemConformanceProvider: Rebuild capability statement notification handled");
-            await _metadataSemaphore.WaitAsync(cancellationToken);
-            try
+
+            if (_builder != null)
             {
-                if (_builder != null)
+                switch (notification.Part)
                 {
-                    switch (notification.Part)
-                    {
-                        case RebuildPart.SearchParameter:
-                            // Update search params;
-                            _builder.SyncSearchParameters();
-                            break;
-
-                        case RebuildPart.Profiles:
-                            // Update supported profiles;
-                            _builder.SyncProfiles(true);
-                            break;
-                    }
+                    case RebuildPart.SearchParameter:
+                        // Update search params;
+                        _builder.SyncSearchParameters();
+                        break;
+
+                    case RebuildPart.Profiles:
+                        // Update supported profiles;
+                        _builder.SyncProfiles(true);
+                        break;
                 }
+            }
 
+            await _metadataSemaphore.WaitAsync(cancellationToken);
```

Comment:
Not sure I understand why the semaphore shouldn't include fetching the data? This reads that many threads may fetch/sync, but must set to null one at a time. 
I feel like the point of this semaphore is more to stop multiple threads syncing at the same time...like a debounce, thoughts?

Reply:
The issue I was running into was that a deadlock situation was being hit. Because I added a refresh for the profiles to the CapabilityStatementBuilder it triggered a RebuildCapabilityStatement notification from the ServiceProvideProfileValidation. So when metadata was being refreshed and using the semaphore the Handle method also tried to acquire it. The metadata couldn't finish its flow until the notification was handled, and the Handle method couldn't complete until it got the semaphore. Only using the semaphore for when the metadata value was being set removed this issue.
It's not a perfect solution, I'm open to suggestions.

Reply:
Ok, I'm understanding the context a little more now, I'm ok with this